chain/epoch-manager/src/validator_selection.rs

```diff
@@ -1265,30 +1239,45 @@ mod tests {
         }
     }
 
+    /// Update the fileds of `EpochConfig` with the given value.
+    macro_rules! update_fields {
+        ($epoch_config:ident, $($field:ident),*) => {
+            $(
+                if let Some(value) = $field {
+                    $epoch_config.$field = value;
+                }
+            )*
+        };
+    }
+
     /// Create EpochConfig, only filling in the fields important for validator selection.
     fn create_epoch_config(
         num_shards: u64,
         num_block_producer_seats: u64,
-        validator_selection_config: ValidatorSelectionConfig,
+        num_chunk_producer_seats: Option<NumSeats>,
+        num_chunk_validator_seats: Option<NumSeats>,
+        num_chunk_only_producer_seats: Option<NumSeats>,
+        minimum_stake_ratio: Option<Ratio<i32>>,
     ) -> EpochConfig {
-        EpochConfig {
-            epoch_length: 10,
-            num_block_producer_seats,
-            num_block_producer_seats_per_shard: vec![num_block_producer_seats; num_shards as usize],
-            avg_hidden_validator_seats_per_shard: vec![0; num_shards as usize],
-            block_producer_kickout_threshold: 0,
-            chunk_producer_kickout_threshold: 0,
-            chunk_validator_only_kickout_threshold: 0,
-            target_validator_mandates_per_shard: 68,
-            validator_max_kickout_stake_perc: 100,
-            online_min_threshold: 0.into(),
-            online_max_threshold: 0.into(),
-            fishermen_threshold: 0,
-            minimum_stake_divisor: 0,
-            protocol_upgrade_stake_threshold: 0.into(),
-            shard_layout: ShardLayout::multi_shard(num_shards, 0),
-            validator_selection_config,
-        }
+        let mut epoch_config = EpochConfig::zero();
+        epoch_config.epoch_length = 10;
+        epoch_config.num_block_producer_seats = num_block_producer_seats;
+        epoch_config.num_block_producer_seats_per_shard =
+            vec![num_block_producer_seats; num_shards as usize];
+        epoch_config.avg_hidden_validator_seats_per_shard = vec![0; num_shards as usize];
+        epoch_config.target_validator_mandates_per_shard = 68;
+        epoch_config.validator_max_kickout_stake_perc = 100;
+        epoch_config.shard_layout = ShardLayout::multi_shard(num_shards, 0);
+
+        update_fields!(
+            epoch_config,
+            num_chunk_producer_seats,
+            num_chunk_validator_seats,
+            num_chunk_only_producer_seats,
+            minimum_stake_ratio
+        );
```

Comment:
It's better to be explicit here, because macros allow to do arbitrarily crazy things, I'd left them for cases where code generation is unavoidable. It's not a big overhead to do couple `if let ...` here.

Also
```
        num_block_producer_seats: u64,
        num_chunk_producer_seats: Option<NumSeats>,
        num_chunk_validator_seats: Option<NumSeats>,
        num_chunk_only_producer_seats: Option<NumSeats>,
```
probably could be derived from only BP and CP seats; COPs shouldn't matter because they are disabled now.

Reply:
Inferring CV and CP form CP and BP needs a bit more attenction. In 2.0 we changed the logic from all nodes being CP to all nodes being CV and in mainnet we made BP = CP.